Shipping rates UI changes: Adjust the logic of prompting for unsaved shipping rates

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #1290 

- Add a new function `hasUnsavedShippingRates` for comparing the saved shipping rates to the one editing via UI.
- Apply `hasUnsavedShippingRates` to check if there are any unsaved shipping rates.

📝  As the implementation is still possible to be adjusted, test cases of the new function will be supplemented with another PR after this one gets merged.

### Screenshots:

💡  To avoid scrolling page all the time, I temporarily removed Hero during the recording

https://user-images.githubusercontent.com/17420811/157431092-ce41d6bc-f1d9-4843-819a-9f80c7a024d5.mp4

### Detailed test instructions:

1. Go to step 2 of the free listings edit page.
2. Make changes to the shipping rates and free shipping threshold.
3. Refresh the current webpage or use other ways to navigate away. There should be a prompt popped-up to ask if you sure want to leave with unsaved changes.
   ![image](https://user-images.githubusercontent.com/17420811/157431829-e952b2f8-598e-4256-861e-cde535e20bb5.png)
4. Revert the changes from step 2.
5. Refresh the current webpage or use other ways to navigate away. It should be able to refresh/leave page directly without prompting.
6. Repeat steps 2-5 with different changes.

### Changelog entry

eason9487 committed:
Compare the key fields to check if there are unsaved shipping rates in free listings edit.

ecgan commented on js/src/edit-free-campaign/hasUnsavedShippingRages.js:
```diff
@@ -0,0 +1,43 @@
+		'options.free_shipping_threshold',
+	];
+
+	const diffRates = differenceWith( rates, savedRates, ( a, b ) => {
+		return isEqual( at( a, paths ), at( b, paths ) );
+	} );
```
❤️ Love this! Short and succinct! I wasn't very familiar with [`differenceWith`](https://lodash.com/docs/4.17.15#differenceWith), but once I learned it from the documentation and then look at this, I got to say this is pretty smart. 😉 👍

ecgan commented on js/src/edit-free-campaign/hasUnsavedShippingRages.js:
```diff
@@ -0,0 +1,43 @@
+		return true;
+	}
+
+	const paths = [
+		'country',
+		'method',
+		'currency',
+		'rate',
+		'options.free_shipping_threshold',
+	];
```
❔ The approach here is to compare specific known paths, i.e. using an "inclusive" approach.

When I first thought about this issue, I was thinking to compare everything in the shipping rate object except `id`, i.e. an "exclusive" approach. 

Do you think if there is any pros and cons to the above? Just curious to hear your opinion. 🤔

eason9487 replied:
In the long term, the exclusive approach should be ideal for shipping data. Not only compatibility toward the future changes, but It's also easier to find out if there are new exclusive properties, like the case in this time, the need to make changes is due to the full exclusivity of the original `isEqual`.

I was a bit hesitant while developing because when the form handles `options.free_shipping_threshold`, the way to clear threshold is to set it to `undefined`, but the API is to unset the property. Two options came to mind at the moment:
1. Adjust the logic in forms - There were other related developments still in progress, in addition to further reducing the flexibility of form handling data.
2. Normalize shipping rates before comparing - But by doing so, it is still an inclusive approach.

So in the short term, I chose inclusive approach first, and we could change it to exclusive approach when someday it's applicable.

ecgan commented on js/src/edit-free-campaign/hasUnsavedShippingRages.js:
```diff
@@ -0,0 +1,43 @@
+ * @param  {Array<ShippingRate>} savedRates The saved shipping rates received from API.
+ * @return {boolean} Whether there are any unsaved shipping changes.
+ */
+export default function hasUnsavedShippingRages( rates, savedRates ) {
```
💅 💅 💅  Please chill, no rages. 🤣 

Note: you got to rename this in your editor, and rename the file name, and also make the necessary changes in your PR description. 😄 

```suggestion
export default function hasUnsavedShippingRates( rates, savedRates ) {
```

eason9487 replied:
Oh no 🤦‍♂️  Thanks for catching this! 😂 

Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/1307/commits/d38dab59344616eacccc48422473eee434be1421.

ecgan commented on js/src/edit-free-campaign/hasUnsavedShippingRages.js:
```diff
@@ -0,0 +1,43 @@
+	];
+
+	const diffRates = differenceWith( rates, savedRates, ( a, b ) => {
+		return isEqual( at( a, paths ), at( b, paths ) );
```
💅 Just a related thought that I had when I was thinking about this issue, we could probably refactor this into a new function like `isEqualShippingRate (a, b)`, then `paths` will be inside that `isEqualShippingRate` function, and then this line will be just `return isEqualShippingRate( a, b );`.

Note: we also already have a function called `isSameShippingRate`. Thinking we might also get into some confusions later. 😅

eason9487 replied:
Thanks for the idea!

I lean toward extracting this when it needs to be reused, then it would not be confused with `isSameShippingRate` before that.

ecgan reviewed (approved):
Tested, working good. I especially love the short and succinct code using lodash `differenceWith` function. It fixes the issues mentioned in https://github.com/woocommerce/google-listings-and-ads/issues/1290. I left a 💅 💅 💅  comment below because of wrong naming (triple emoji to make sure you really sees it 😄 ). I trust that you will fix it, so I'm approving in advance. 😉 

Thanks so much for this PR! 🙂 👍

eason9487 committed:
Fix the naming typos of `hasUnsavedShippingRates` function, file and path.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1307#discussion_r822944729
